File: 2024/day16/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# Day 16: Reindeer Maze - Weighted shortest path with turn costs

# Directions: 0=East, 1=South, 2=West, 3=North
my @DX = (1, 0, -1, 0);
my @DY = (0, 1, 0, -1);

sub parse_input {
    my ($text) = @_;
    my @grid = map { [split //] } split /\n/, $text;
    my ($start, $end);

    for my $y (0 .. $#grid) {
        for my $x (0 .. $#{$grid[$y]}) {
            if ($grid[$y][$x] eq 'S') {
                $start = [$x, $y];
            } elsif ($grid[$y][$x] eq 'E') {
                $end = [$x, $y];
            }
        }
    }

    return (\@grid, $start, $end);
}

sub dijkstra_forward {
    my ($grid, $start) = @_;
    my @pq = ([0, $start->[0], $start->[1], 0]);  # Start facing East
    my %dist;

    while (@pq) {
        # Extract minimum by cost
        @pq = sort { $a->[0] <=> $b->[0] } @pq;
        my ($cost, $x, $y, $d) = @{shift @pq};

        my $state = "$x,$y,$d";
        next if exists $dist{$state};
        $dist{$state} = $cost;

        # Move forward
        my $nx = $x + $DX[$d];
        my $ny = $y + $DY[$d];
        if ($ny >= 0 && $ny < @$grid && $nx >= 0 && $nx < @{$grid->[$ny]} &&
            $grid->[$ny][$nx] ne '#') {
            push @pq, [$cost + 1, $nx, $ny, $d];
        }

        # Turn left/right
        push @pq, [$cost + 1000, $x, $y, ($d - 1) % 4];
        push @pq, [$cost + 1000, $x, $y, ($d + 1) % 4];
    }

    return \%dist;
}

sub dijkstra_backward {
    my ($grid, $end) = @_;
    # At end, we can arrive facing any direction
    my @pq = map { [0, $end->[0], $end->[1], $_] } (0..3);
    my %dist;

    while (@pq) {
        # Extract minimum by cost
        @pq = sort { $a->[0] <=> $b->[0] } @pq;
        my ($cost, $x, $y, $d) = @{shift @pq};

        my $state = "$x,$y,$d";
        next if exists $dist{$state};
        $dist{$state} = $cost;

        # Reverse of "move forward": come from behind
        my $px = $x - $DX[$d];
        my $py = $y - $DY[$d];
        if ($py >= 0 && $py < @$grid && $px >= 0 && $px < @{$grid->[$py]} &&
            $grid->[$py][$px] ne '#') {
            push @pq, [$cost + 1, $px, $py, $d];
        }

        # Reverse of turn: came from same position with different direction
        push @pq, [$cost + 1000, $x, $y, ($d - 1) % 4];
        push @pq, [$cost + 1000, $x, $y, ($d + 1) % 4];
    }

    return \%dist;
}

sub part1 {
    my ($grid, $start, $end) = @_;
    my $dist = dijkstra_forward($grid, $start);

    my $min_cost = 1e100;
    for my $d (0..3) {
        my $state = "$end->[0],$end->[1],$d";
        if (exists $dist->{$state} && $dist->{$state} < $min_cost) {
            $min_cost = $dist->{$state};
        }
    }

    return $min_cost;
}

sub part2 {
    my ($grid, $start, $end, $best_score) = @_;
    my $dist_from_start = dijkstra_forward($grid, $start);
    my $dist_to_end = dijkstra_backward($grid, $end);

    my %tiles_on_best_path;

    for my $y (0 .. $#{$grid}) {
        for my $x (0 .. $#{$grid->[$y]}) {
            next if $grid->[$y][$x] eq '#';

            # Check if this tile is on any optimal path
            for my $d (0..3) {
                my $state = "$x,$y,$d";
                my $from_start = $dist_from_start->{$state} // 1e100;
                my $to_end = $dist_to_end->{$state} // 1e100;

                if ($from_start + $to_end == $best_score) {
                    $tiles_on_best_path{"$x,$y"} = 1;
                    last;
                }
            }
        }
    }

    return scalar keys %tiles_on_best_path;
}

sub main {
    # Read input
    my $input_path = '../input.txt';
    open my $fh, '<', $input_path or die "Cannot open $input_path: $!";
    my $text = do { local $/; <$fh> };
    close $fh;

    chomp $text;
    my ($grid, $start, $end) = parse_input($text);

    my $answer1 = part1($grid, $start, $end);
    say "Part 1: $answer1";

    my $answer2 = part2($grid, $start, $end, $answer1);
    say "Part 2: $answer2";
}

main();

File: 2024/day16/ruby/solution.rb
#!/usr/bin/env ruby
# Day 16: Reindeer Maze - Weighted shortest path with turn costs

require 'set'

def parse_input(text)
  grid = text.strip.split("\n").map(&:chars)
  start_pos = nil
  end_pos = nil

  grid.each_with_index do |row, y|
    row.each_with_index do |cell, x|
      start_pos = [x, y] if cell == 'S'
      end_pos = [x, y] if cell == 'E'
    end
  end

  [grid, start_pos, end_pos]
end

# Directions: 0=East, 1=South, 2=West, 3=North
DX = [1, 0, -1, 0]
DY = [0, 1, 0, -1]

def dijkstra_forward(grid, start)
  # Priority queue: [cost, x, y, direction]
  # Ruby doesn't have built-in heap, so we'll use a sorted array
  pq = [[0, start[0], start[1], 0]] # Start facing East
  dist = {}

  until pq.empty?
    cost, x, y, d = pq.shift

    state = [x, y, d]
    next if dist.key?(state)
    dist[state] = cost

    # Move forward
    nx = x + DX[d]
    ny = y + DY[d]
    if ny >= 0 && ny < grid.length && nx >= 0 && nx < grid[0].length && grid[ny][nx] != '#'
      insert_sorted(pq, [cost + 1, nx, ny, d])
    end

    # Turn left and right
    insert_sorted(pq, [cost + 1000, x, y, (d - 1) % 4])
    insert_sorted(pq, [cost + 1000, x, y, (d + 1) % 4])
  end

  dist
end

def dijkstra_backward(grid, end_pos)
  # Run Dijkstra backward from end (all directions at end have cost 0)
  pq = (0..3).map { |d| [0, end_pos[0], end_pos[1], d] }
  dist = {}

  until pq.empty?
    cost, x, y, d = pq.shift

    state = [x, y, d]
    next if dist.key?(state)
    dist[state] = cost

    # Reverse of "move forward": come from behind
    px = x - DX[d]
    py = y - DY[d]
    if py >= 0 && py < grid.length && px >= 0 && px < grid[0].length && grid[py][px] != '#'
      insert_sorted(pq, [cost + 1, px, py, d])
    end

    # Reverse of turn: came from same position with different direction
    insert_sorted(pq, [cost + 1000, x, y, (d - 1) % 4])
    insert_sorted(pq, [cost + 1000, x, y, (d + 1) % 4])
  end

  dist
end

def insert_sorted(arr, item)
  # Binary search insertion to maintain sorted order by cost
  left = 0
  right = arr.length

  while left < right
    mid = (left + right) / 2
    if arr[mid][0] < item[0]
      left = mid + 1
    else
      right = mid
    end
  end

  arr.insert(left, item)
end

def part1(grid, start, end_pos)
  dist = dijkstra_forward(grid, start)
  (0..3).map { |d| dist[[end_pos[0], end_pos[1], d]] || Float::INFINITY }.min
end

def part2(grid, start, end_pos, best_score)
  dist_from_start = dijkstra_forward(grid, start)
  dist_to_end = dijkstra_backward(grid, end_pos)

  tiles_on_best_path = Set.new

  grid.each_with_index do |row, y|
    row.each_with_index do |cell, x|
      next if cell == '#'

      # Check if this tile is on any optimal path
      (0..3).each do |d|
        state = [x, y, d]
        from_start = dist_from_start[state] || Float::INFINITY
        to_end = dist_to_end[state] || Float::INFINITY

        if from_start + to_end == best_score
          tiles_on_best_path.add([x, y])
          break
        end
      end
    end
  end

  tiles_on_best_path.size
end

def main
  input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
  text = File.read(input_path)
  grid, start, end_pos = parse_input(text)

  answer1 = part1(grid, start, end_pos)
  puts "Part 1: #{answer1}"

  answer2 = part2(grid, start, end_pos, answer1)
  puts "Part 2: #{answer2}"
end

main if __FILE__ == $PROGRAM_NAME
